datafusion/common/src/tree_node.rs

```diff
@@ -204,27 +176,64 @@ pub trait TreeNode: Sized {
     ///
     /// The nodes are visited using the following order:
     /// ```text
-    /// f_down(ParentNode)
-    /// f_down(ChildNode1)
-    /// f_up(ChildNode1)
-    /// f_down(ChildNode2)
-    /// f_up(ChildNode2)
-    /// f_up(ParentNode)
+    /// TreeNodeRewriter::f_down(ParentNode)
+    /// TreeNodeRewriter::f_down(ChildNode1)
+    /// TreeNodeRewriter::f_up(ChildNode1)
+    /// TreeNodeRewriter::f_down(ChildNode2)
+    /// TreeNodeRewriter::f_up(ChildNode2)
+    /// TreeNodeRewriter::f_up(ParentNode)
     /// ```
     ///
     /// See [`TreeNodeRecursion`] for more details on how the traversal can be controlled.
     ///
-    /// If `f_down` or `f_up` returns [`Err`], recursion is stopped immediately.
-    fn transform<FD, FU>(
+    /// If [`TreeNodeRewriter::f_down()`] or [`TreeNodeRewriter::f_up()`] returns [`Err`],
+    /// recursion is stopped immediately.
+    fn rewrite<R: TreeNodeRewriter<Node = Self>>(
         self,
-        f_down: &mut FD,
-        f_up: &mut FU,
-    ) -> Result<Transformed<Self>>
-    where
-        FD: FnMut(Self) -> Result<Transformed<Self>>,
-        FU: FnMut(Self) -> Result<Transformed<Self>>,
-    {
-        handle_transform_recursion!(f_down(self), |c| c.transform(f_down, f_up), f_up)
+        rewriter: &mut R,
+    ) -> Result<Transformed<Self>> {
+        let pre_visited = rewriter.f_down(self)?;
+        match pre_visited.tnr {
+            TreeNodeRecursion::Continue => {
+                let with_updated_children = pre_visited
+                    .data
+                    .map_children(|c| c.rewrite(rewriter))?
+                    .try_transform_node_with(
+                        |n| rewriter.f_up(n),
+                        Some(TreeNodeRecursion::Jump),
+                    )?;
+                Ok(Transformed {
+                    transformed: with_updated_children.transformed
+                        || pre_visited.transformed,
+                    ..with_updated_children
+                })
+            }
+            TreeNodeRecursion::Jump => {
+                let pre_visited_transformed = pre_visited.transformed;
+                let post_visited = rewriter.f_up(pre_visited.data)?;
+
+                Ok(Transformed {
+                    tnr: TreeNodeRecursion::Continue,
+                    transformed: post_visited.transformed || pre_visited_transformed,
+                    data: post_visited.data,
+                })
+            }
+            TreeNodeRecursion::Stop => Ok(pre_visited),
+        }
+    }
+
+    /// Applies `f` to the node and its children. `f` is applied in a preoder way,
```

Comment:
Any reason to move `apply()` here?

Reply:
I just wanted to keep `visit` and `rewrite` together. `apply` also does the same traversal of `transform_down`. But you can move it anywhere as you wish.